### PR TITLE
Improve testing with Ibis

### PR DESCRIPTION
**Pull request by Tishj**
This PR adds classes that reduces the amount of boilerplate needed to test our results against/with Ibis

## Tester Interface
`IbisDuckDBTester` is an interface that defines the skeleton of the tester.
To derive from an IbisDuckDBTester, the only thing that needs to be created is one method:
```py
        def generate_relation(self, expr) -> duckdb.DuckDBPyRelation:
```
This method is used to transform the Ibis expression to an equivalent DuckDBPyRelation.

## API
To set up the tester, you provide a list of sql queries, they are run sequentially at start up.

To interact with the tester, a single `test` method is exposed:
```py
	def test(self, expression_producer, *args):
```
It's arguments are:
- `expression_producer`
A function that produces an Ibis expression, given an ibis connection, for example:
```py
def extract_component(ibis_db, named_component):
	tbl = ibis_db.table('tbl')
	expr = tbl[getattr(tbl.d.time(), named_component)().cast('int64')]
	return expr
```
- `*args`
Any arguments required by the `expression_producer`, these will be forwarded by the tester

## Implementations
- `SQLIbisDuckDBTester`
This tester uses `ibis.to_sql(expr)` to convert the expression to sql, and then uses `duckdb.sql` to create a relation out of it.

- `SubstraitIbisDuckDBTester`
This tester uses `SubstraitCompiler` to generate the substrait from an expression, which we then convert to JSON, and finally we use `duckdb.from_substrait_json` to create a relation from the substrait plan.

## Coverage
For ease of use and to maximize coverage, this PR also adds a `CombinedIbisDuckDBTester` which bundles the existing implementations and `test` forwards the calls to the internal testers.

## Technical details
The internal implementation (summarized) is as follows:
We create a persistent duckdb database
populate it using the initialization queries.
crearte a connection to the duckdb database with ibis

Every time `test` is invoked:
create an expression on it (using the `expression_producer`)
create a duckdb relation from the ibis expression (using the *virtual* `generate_relation` method)
execute both the relation and the expression, outputting to pyarrow - asserting that they produce the same result

### CODE REVIEW

**Comment by Tishj** [comment]
So I don't forget, this needs a follow-up PR to add `duckdb` -> `substrait` -> `ibis` verification.
This PR is for `ibis` -> `substrait` -> `duckdb`

File: test/python/ibis/test_extract_timestamp.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+			#"epoch_seconds",
+			"month",
+			#"quarter",
+			#"week_of_year",
+			"year",
+			#"day_of_week", # not sure if this one should be in this list at all
+			#"hour",
+			#"millisecond",
+			#"minute",
+			#"second"
```
Comment (pdet) [inline test/python/ibis/test_extract_timestamp.py:39]: Should we break these into different tests, and the ones not supported have pytest skip? I think it's cleaner

File: test/python/ibis_duckdb_tester.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	def test(self, expression_producer, *args):
+		expr = expression_producer(self.ibis_con, *args)
+		relation = self.generate_relation(expr.unbind())
+
+		# Verify that the expressions produce the same result
+		res = expr.to_pyarrow()
+		duck_res = relation.arrow()
+		#print(res)
+		#print(duck_res)
```
Comment (pdet) [inline test/python/ibis_duckdb_tester.py:34]: gunk?

**Comment by gforsyth** [comment]
> So I don't forget, this needs a follow-up PR to add duckdb -> substrait -> ibis verification.

A heads-up that the substrait decompiler in ibis-substrait is out-of-date and not particularly full-featured

**Comment by Tishj** [comment]
Seems something happened to timestamp extract support since I made this PR?

```py
duck = duckdb.connect(db_path)

duck.execute("""
	create table tbl(date timestamp)
""")
duck.execute("""
	insert into tbl values ('2021/09/21 12:02:21'::TIMESTAMP)
""")
duck.close()

ibis_db = ibis.connect(f'duckdb://{db_path}', read_only='True')
tbl = ibis_db.table('tbl')
expr = tbl[getattr(tbl.date, 'day')().cast('int64')]
expr = expr.unbind()

import ibis_substrait.compiler.core as SubstraitCompiler

compiler = SubstraitCompiler.SubstraitCompiler()
from google.protobuf import json_format

proto = compiler.compile(expr)
```

```
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/core.py", line 79, in function_id
    function_extension = self.function_extensions[op_name]
KeyError: 'extract'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/thijs/DuckDBLabs/substrait/../duckdb/tmp/ibis_test.py", line 35, in <module>
    proto = compiler.compile(expr)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/core.py", line 193, in compile
    input=translate(expr.op(), compiler=self, **kwargs),
  File "/opt/homebrew/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 888, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/translate.py", line 868, in selection
    expressions=[
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/translate.py", line 869, in <listcomp>
    translate(
  File "/opt/homebrew/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 888, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/translate.py", line 247, in _expr
    return translate(expr.op(), compiler=compiler, **kwargs)
  File "/opt/homebrew/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 888, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/translate.py", line 1188, in _cast
    input=translate(op.arg, compiler=compiler, **kwargs),
  File "/opt/homebrew/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 888, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/translate.py", line 1209, in _extractdatefield
    function_reference=compiler.function_id(op),
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/core.py", line 83, in function_id
    ] = self.extension_lookup(op_name, op)
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/core.py", line 109, in extension_lookup
    [
  File "/opt/homebrew/lib/python3.9/site-packages/ibis_substrait/compiler/core.py", line 110, in <listcomp>
    IBIS_SUBSTRAIT_TYPE_MAPPING[arg.op().output_dtype.name]  # type: ignore
KeyError: 'Timestamp'
```

**Comment by pdet** [comment]
Thanks for the work Thijs, I guess the spurious R fail is still going on :(

I just have two small requests:
1) Can you add a `xfail` to the tests that are failing due to lack of ibis support? It's easier for us to switch them back when it gets implemented.

2) Can you also add a comment with a short list of the missing components so we can expose them to Gil/Ian?

**Comment by pdet** [comment]
@Tishj are you still working on this?

**Pull request closed without merging**